hr_contract_cae/models/hr_contract.py

```diff
@@ -19,20 +144,162 @@ class Contract(models.Model):
             self.duration = rd.months + rd.years * 12
+
+    @api.multi
+    @api.depends("contract_group_contract_ids", "type_id")
+    def _compute_type_count(self):
+        for contract in self:
+            type_count = 0
+            for c in contract.contract_group_contract_ids:
+                if c.type_id == contract.type_id:
+                    type_count += 1
+            contract.type_count = type_count
+
+    @api.multi
+    def _compute_attachment_number(self):
+        read_group_res = self.env["ir.attachment"].read_group(
+            [("res_model", "=", "hr.contract"), ("res_id", "in", self.ids)],
+            ["res_id"],
+            ["res_id"],
+        )
+        attach_data = {
+            res["res_id"]: res["res_id_count"] for res in read_group_res
+        }
+        for contract in self:
+            contract.attachment_number = attach_data.get(contract.id, 0)
+
+    @api.multi
+    def action_get_attachment_tree_view(self):
+        self.ensure_one()
+        attachment_action = self.env.ref("base.action_attachment")
+        action = attachment_action.read()[0]
+        action["context"] = {
+            "default_res_model": self._name,
+            "default_res_id": self.ids[0],
+        }
+        action["domain"] = str(
+            ["&", ("res_model", "=", self._name), ("res_id", "in", self.ids)]
+        )
+        return action
+
+    #  Note: this check must run both when creating a new amendment
+    #  and when changing the type_id of an amendment
+    @api.multi
+    @api.constrains("type_count", "type_id")
+    def check_type_count(self):
+        for contract in self:
+            if (
+                contract.type_id.max_usage
+                and contract.type_count > contract.type_id.max_usage
+            ):
+                raise ValidationError(
+                    _(
+                        "The maximum amount of %s contracts "
+                        "of type '%s' has been reached"
+                        % (contract.type_id.max_usage, contract.type_id.name)
+                    )
+                )
+
+    def _compute_initial(self):
+        for contract in self:
+            contract.initial_contract_id = contract.get_initial()
+
+    def _compute_parent(self):
+        for contract in self:
+            contract.parent_contract_id = contract.get_parent()
+
+    def _compute_child(self):
+        for contract in self:
+            contract.child_contract_id = contract.get_child()
+
+    def _compute_latest(self):
+        for contract in self:
+            contract.latest_contract_id = contract.get_latest()
+
+    def get_initial(self):
+        return self.contract_group_id.get_initial()
+
+    def get_parent(self):
+        return self.contract_group_id.get_parent(self.amendment_index)
+
+    def get_child(self):
+        return self.contract_group_id.get_child(self.amendment_index)
+
+    def get_latest(self):
+        return self.contract_group_id.get_latest()
```

Comment:
There are too many layers here :-)

Reply:
This layer could go indeed and be implemented in the `_compute` functions. These getters are of no use since the values are stored in the fields...